src/fetchers/tarball-fetcher.js

```diff
@@ -275,24 +290,43 @@ export default class TarballFetcher extends BaseFetcher {
               integrityValidateStream.pipe(fs.createWriteStream(tarballCachePath)).on('error', reject);
             }
 
-            integrityValidateStream.pipe(extractorStream).on('error', reject);
+            integrityValidateStream.pipe(extractorStream).on('error', async err => {
+              try {
+                // Cleanup before retrying
+                await this.removeDownloadedFiles();
+                if (this.validateError && this.validateError.code == 'EBADSIZE') {
+                  // Content-Length did not match the body size. Retry the request.
+                  if (!queueForRetry(err.message)) {
+                    reject(err);
+                  }
+                  return;
+                }
+                reject(err);
+              } catch (err2) {
+                // Not expected, but handle rather be safe.
+                reject(err2);
+              }
+            });
           },
         },
         this.packageName,
       );
     } catch (err) {
-      const tarballMirrorPath = this.getTarballMirrorPath();
-      const tarballCachePath = this.getTarballCachePath();
+      await this.removeDownloadedFiles();
+      throw err;
+    }
+  }
 
-      if (tarballMirrorPath && (await fsUtil.exists(tarballMirrorPath))) {
-        await fsUtil.unlink(tarballMirrorPath);
-      }
+  async removeDownloadedFiles(): Promise<void> {
+    const tarballMirrorPath = this.getTarballMirrorPath();
+    const tarballCachePath = this.getTarballCachePath();
 
-      if (tarballCachePath && (await fsUtil.exists(tarballCachePath))) {
-        await fsUtil.unlink(tarballCachePath);
-      }
+    if (tarballMirrorPath && (await fsUtil.exists(tarballMirrorPath))) {
```

Comment:
I know this is existing code but we can probably use this as an opportunity to "fix" it: it is discouraged to use the async `exists` method to check for a file and try operating on it.

It'd probably be better to refactor this to something like:
```js
try {
return await Promise.all([
  tarballMirrorPath ? fsUtil.unlink(tarballMirrorPath) : Promise.resolve(true),
  tarballCachePath ? fsUtil.unlink(tarballCachePath) : Promise.resolve(true),
]);
} catch (err) {
  // Ignore if file does not exist
  if (err.code !== 'ENOENT') {
    raise err;
  }
}
```